param/serializer.py

```diff
@@ -0,0 +1,178 @@
+"""
+Classes used to support string serialization of Parameters and
+Parameterized objects.
+"""
+
+import json
+import datetime as dt
+
+class UnserializableException(Exception):
+    pass
+
+class UnsafeserializableException(Exception):
+    pass
+
+def JSONNullable(json_type):
+    "Express a JSON schema type as nullable to easily support Parameters that allow_None"
+    return { "anyOf": [ json_type, { "type": "null"}] }
+
+
+
+class Serialization(object):
+    """
+    Base class used to implement different types of serialization.
+    """
+
+    @classmethod
+    def schema(cls, pobj):
+        raise NotImplementedError
+
+    @classmethod
+    def serialize_parameters(cls, pobj):
+        raise NotImplementedError
+
+
+
+class JSONSerialization(Serialization):
+    """
+    Class responsible for specifying JSON serialization, deserialization
+    and JSON schemas for Parameters and Parameterized classes and
+    objects.
+    """
+
+    unserializable_parameter_types = ['Callable']
+
+    json_schema_literal_types = {int:'integer', float:'number', str:'string'}
```

Comment:
param.String schema should include the optional regex attribute: https://json-schema.org/understanding-json-schema/reference/regular_expressions.html

Reply:
Would be nice but JavaScript and Python regular expressions are not compatible enough for this to be worth attempting.